### PR TITLE
Added the ability to update recommendations

### LINKED ISSUE #17: Need ability to Update a recommendation
**As a** manager
 **I need** to create a service to edit resource information in Database
 **So that** I can maintain a up-to-date database of resources  
   
 ### Details and Assumptions
 * resource information is put in a file
   
 ### Acceptance Criteria  
   
 ```gherkin
 Given to edit information certain information regarding a resource
 When a API call is made to edit 
 Then the database file is edited with new information
 ```

### PR DESCRIPTION
**Pull request by calboy404**

### CODE REVIEW

**Comment by calboy404** [comment]
Added the ability to update recommendations

File: service/models.py [hunk 1]
```diff
@@ -50,6 +50,18 @@ def create(self):
...
+     def update(self):
+        """
+        Updates a Recommendation to the database
+        """
+        logger.info("Saving %s", self.name)
+        if not self.id:
+            raise DataValidationError("Update called with empty ID field")
+        if type(self.id) is not int:
+            raise DataValidationError("Update called with non-integer ID field")
+        if type(self.price) is not int:
```
Comment (aianurag09) [inline service/models.py]: There is no price in the Recomendation Model, change the code according to the attributes of Recommendation model i.e src_prod_id etc.

File: tests/test_models.py [hunk 2]
```diff
@@ -112,3 +112,31 @@ def test_add_a_recommendation(self):
         self.assertEqual(recommendation.id, 1)
         recommendations = Recommendation.all()
         self.assertEqual(len(recommendations), 1)
+        
+    def test_update_a_recommendation(self):
+        """Update a item"""
+        recommendation = Recommendation(id=1, name="iPhone", category="phone", price=100)
```
Comment (aianurag09) [inline tests/test_models.py]: The attributes are change refer to attributes in model.py

File: tests/test_models.py [hunk 3]
```diff
@@ -112,3 +112,31 @@ def test_add_a_recommendation(self):
...
+        # Fetch it back and make sure the id hasn't changed
+        # but the data did change
+        recommendations = Recommendation.all()
+        self.assertEqual(len(recommendations), 1)
+        self.assertEqual(recommendations[0].id, 1)
+        self.assertEqual(recommendations[0].category, "laptop")
+
+    def test_update_a_recommendation_validation_error(self):
+        """Update a item Validation Error"""
+        recommendation = Recommendation(id=1, name="iPhone", category="Phone", price="100")
```
Comment (aianurag09) [inline tests/test_models.py]: same change the attributes and check with the attributes in modes.phy

**Comment by aianurag09** [comment]
@calboy404 you didn;t change the code according to the attributes, refer to other pull requests and models.py and change accordingly

**Review by aianurag09 — APPROVED** [review]

**Comment by aianurag09** [comment]
Done @calboy404

File: service/models.py [hunk 4]
```diff
@@ -50,6 +50,20 @@ def create(self):
         self.id = None  # id must be none to generate next primary key
         db.session.add(self)
         db.session.commit()
+        
+     def update(self):
+        """
+        Updates a Recommendation to the database
+        """
+        logger.info("Saving %s", self.name)
```
Comment (changtk07) [inline service/models.py]: There's no `name` attribute, use `id`

File: service/models.py [hunk 5]
```diff
@@ -50,6 +50,20 @@ def create(self):
...
+        if not self.id:
+            raise DataValidationError("Update called with empty ID field")
+        if type(self.id) is not int:
```
Comment (changtk07) [inline service/models.py]: Line 61-66, no need to check these

File: service/routes.py [hunk 6]
```diff
@@ -69,7 +69,23 @@ def create_Recommendation():
...
+    app.logger.info("Request to update a recommendation with id: %s", item_id)
+    check_content_type("application/json")
+    recommendation = Recommendation.find(item_id)
+    if not recommendation:
+        raise NotFound("recommendation with id '{}' was not found.".format(item_id))
+    recommendation.deserialize(request.get_json())
+    recommendation.id = item_id
+    recommendation.update()
 
+    app.logger.info("recommendation with ID [%s] updated.", product.id)
```
Comment (changtk07) [inline service/routes.py]: `product` -> `recommendation`

File: tests/test_models.py [hunk 7]
```diff
@@ -112,3 +112,32 @@ def test_add_a_recommendation(self):
         self.assertEqual(recommendation.id, 1)
         recommendations = Recommendation.all()
         self.assertEqual(len(recommendations), 1)
+        
+    def test_update_a_recommendation(self):
+        """Update an item"""
```
Comment (changtk07) [inline tests/test_models.py]: item -> recommendation

File: tests/test_models.py [hunk 8]
```diff
@@ -112,3 +112,32 @@ def test_add_a_recommendation(self):
...
+        self.assertEqual(recommendations[0].rec_product_id, 400)
+        self.assertEqual(recommendations[0].type, "UP_SELL")
+
+    def test_update_a_recommendation_validation_error(self):
+        """Update a item Validation Error"""
+        recommendation = Recommendation(id=1, src_product_id=100, rec_product_id = 200, type= "UP_SELL")
+        self.assertRaises(DataValidationError, recommendation.update)
+        recommendation = Recommendation(src_product_id=100, rec_product_id = 200, type= "UP_SELL")
+        self.assertRaises(DataValidationError, recommendation.update)
+        recommendation = Recommendation(id=1, src_product_id=100, rec_product_id = 200, type= "UP_SELL")
```
Comment (changtk07) [inline tests/test_models.py:167]: What's the difference of this test and line 138?

File: tests/test_routes.py [hunk 9]
```diff
@@ -113,3 +113,25 @@ def test_create_recommendation_no_content_type(self):
...
+        test_recommendation = RecommendationFactory()
+        resp = self.app.post(
+            BASE_URL, json=test_recommendation.serialize(), content_type=CONTENT_TYPE_JSON
+        )
+        self.assertEqual(resp.status_code, status.HTTP_201_CREATED)
+
+        # update the recommendation
+        new_recommendation = resp.get_json()
+        logging.debug(new_recommendation)
+        new_recommendation["category"] = "unknown"
```
Comment (changtk07) [inline tests/test_routes.py]: no `category` attribute

File: tests/test_routes.py [hunk 10]
```diff
@@ -113,3 +113,25 @@ def test_create_recommendation_no_content_type(self):
...
+        resp = self.app.put(
+            "/recommendations/{}".format(new_recommendation["id"]),
+            json=new_recommendation,
+            content_type=CONTENT_TYPE_JSON,
+        )
+        self.assertEqual(resp.status_code, status.HTTP_200_OK)
+        updated_recommendation = resp.get_json()
+        self.assertEqual(updated_recommendation["category"], "unknown")
```
Comment (changtk07) [inline tests/test_routes.py]: no `category` attribute

**Review by changtk07 — CHANGES_REQUESTED** [review]
Please run nosetests, these error won't even pass the tests.
Please test before your pull request

File: service/models.py [hunk 11]
```diff
@@ -50,6 +50,20 @@ def create(self):
         self.id = None  # id must be none to generate next primary key
         db.session.add(self)
         db.session.commit()
+        
+     def update(self):
```
Comment (changtk07) [inline service/models.py]: Wrong indentation

File: tests/test_routes.py [hunk 12]
```diff
@@ -113,3 +113,25 @@ def test_create_recommendation_no_content_type(self):
         """Create a recommendation with no content type"""
         resp = self.app.post(BASE_URL)
         self.assertEqual(resp.status_code, status.HTTP_415_UNSUPPORTED_MEDIA_TYPE)
+        
+     def test_update_recommendatio(self):
```
Comment (changtk07) [inline tests/test_routes.py]: Wrong indentation

**Review by changtk07 — CHANGES_REQUESTED** [review]

**Review by changtk07 — APPROVED** [review]

**Pull request merged**